### PR TITLE
Fixed read_csv with CategoricalDtype with boolean categories (20498)

### LINKED ISSUE #20498: Force boolean column to category while reading a csv
#### Problem description

Reading a csv file with dtypes specified in a dictionary produces NaNs for boolean columns read as category. It seems the output of dtypes changes from version 0.20 to 0.21 so that the below code produces NaNs for the second column. 

This might not really be a bug but it seems quite unintuitive to me as I would expect that output of df.dtypes should be useable to read it correctly again and I would expect the interpretation of booleans as a category as valid.

#### Code to reproduce the issue
```python
b = {'a':[5,4,3,2], 'b':[True, False, True, True], , 'c':['A', 'B', 'A', 'C']}
df = pd.DataFrame(b)

df['b'] = df['b'].astype('category')
df['c'] = df['c'].astype('category')
dtypes_dict = df.dtypes.to_dict()
df.to_csv("data.csv", index=False)

df2 = pd.read_csv("data.csv", dtype=dtypes_dict)
print(df2)
```

#### Produced Output
```
   a    b  c
0  5  NaN  A
1  4  NaN  B
2  3  NaN  A
3  2  NaN  C
```

#### Expected Output
```
   a    b  c
0  5  True  A
1  4  False  B
2  3  True  A
3  2  True  C
```
#### Output of df.dtypes version 0.22
```{'a': dtype('int64'), 'b': CategoricalDtype(categories=[False, True], ordered=False), 'c': CategoricalDtype(categories=['A', 'B', 'C'], ordered=False)}```

#### Output of df.dtypes version 0.20
```{'a': dtype('int64'), 'b': category, 'c': category}```

#### Version
Pands version 0.22
<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.8.6-040806-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_IE.UTF-8
LOCALE: en_IE.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.2
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.0
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by TomAugspurger**
- [x] closes #20498
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/categorical.py
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: Cython experts, is there a way to share a declaration with a non cython file?

We have

```
cdef object _true_values = [b'True', b'TRUE', b'true']
```

changing that do a `cpdef` didn't work for some reason.

(we need this, since the PythonParser's `true_values` can apparently be none.

Comment (jreback) [reply pandas/core/arrays/categorical.py]: you need to remove the ``cdef`` then its importable just like anything else

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py]: this does NOT belong here at all, you are mixing up parsing with setting the categories. you can do this via a separate function *inside* parser.pyx

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py
Comment (TomAugspurger) [reply pandas/core/arrays/categorical.py]: I think there were issues with that (either nogil, or having to share code with the Python parser).

Comment (jreback) [reply pandas/core/arrays/categorical.py]: well it certainly doesn’t belong here
this is conflating parsing with categories

Comment (TomAugspurger) [reply pandas/core/arrays/categorical.py]: I see now that this is called outside a nogil block, so a standalone method for that should be fine. If you want to open an issue though, because I don't think it should hold up the release.

Comment (jreback) [reply pandas/core/arrays/categorical.py]: this isn't necessary for the RC. this needs to be fixed in this PR.

**Comment by TomAugspurger** [comment]
I don't think I'll have time before the RC, and we're simply fixing a bug
in the existing code.

I'd like to keep the just merge bug fixes with the RC before the final, so
your call if you want to push to 0.23.1.

**Comment by jreback** [comment]
then push to 0.23.1 (alternatively as @jorisvandenbossche mentoined, can make a 0.23rc milestone tag)

File: doc/source/whatsnew/v0.23.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: can you move to 0.23.1

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py]: versionadded?

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py]: you could *maybe* enhance ``to_numeric`` to take a ``downcasts='bool'`` to do this?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
@TomAugspurger do you have time to look at this one? (I don't know how far off it is)

**Comment by TomAugspurger** [comment]
Possibly. If you can provide a unit test I can include it or you can make a followup PR including it.

File: pandas/tests/categorical/test_constructors.py
Comment (jreback) [inline pandas/tests/categorical/test_constructors.py]: would be ok with removing these tests (as I think you have them covered directly in the parser)

File: doc/source/whatsnew/v0.24.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: is this enough from a user POV? need a small example?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
IIRC this was fine, can you rebase

**Comment by jreback** [comment]
this looks good; haven’t looked at what is failing

**Comment by jreback** [comment]
@TomAugspurger can you merge master

**Comment by jreback** [comment]
@gfyoung if you'd like to rebase this, maybe can get this in

**Comment by gfyoung** [comment]
Yikes!  Looks like this one got lost in the myriad of other PR's.  Sure thing.

**Comment by gfyoung** [comment]
@jreback  @TomAugspurger : Rebased this PR.  Wanted to note a couple of things (also helps to address many of the outstanding comments that were still current):

* I personally was not a fan of moving an entire `Categorical` method into `parsers.pyx`.  That IMO seems to do the kind of conflating that was pushed back against earlier.  I was able to get things to work without making this migration.  At the same time, the "parsing" is relegated to `parsers.pyx` before being passed to `Categorical._from_inferred_categories`.
* I'm okay with having a single `whatsnew` entry line.  I reworded it a little to make things clearer.
* Not sure we need a version added on the docstring, given that `Categorical._from_inferred_categories` is a private method by naming (starts with an underscore).

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks ok to me. @TomAugspurger if you'd have a look.

**Comment by gfyoung** [comment]
@TomAugspurger : Any thoughts on this?  I think we're waiting on your review before merging.

**Comment by TomAugspurger** [comment]
Sorry for the delay, thanks for rebasing and fixing things up.

**Pull request merged**
